### PR TITLE
fix #117901: support #include, change #define behavior, support <global>

### PR DESCRIPTION
**Pull request by hpfmn**
for more information see https://musescore.org/en/node/117901

### CODE REVIEW

File: zerberus/sfz.cpp [hunk 1]
```diff
@@ -529,8 +555,32 @@ bool ZInstrument::loadSfz(const QString& s)
...
+                              newFilename.replace(define.first, define.second);
+                              }
+
+                        fileStack.push(f);
+                        QFile *newFile = new QFile(path + "/" + newFilename);
+
+                        f = newFile;
+                        if (!f->open(QIODevice::ReadOnly)) {
+                              qDebug("ZInstrument: cannot load %s", qPrintable(path + "/" + newFilename));
+                              return false;
```
Comment (lasconic) [inline zerberus/sfz.cpp]: f is not deleted here right?

File: zerberus/sfz.cpp [hunk 2]
```diff
@@ -41,7 +42,7 @@ void SfzControl::init()
       {
       defaultPath.clear();
       octave_offset = 0;
-      defines.clear();
+      //defines.clear();
```
Comment (lasconic) [inline zerberus/sfz.cpp]: If it's not needed, let's delete it and not comment it. It will raise questions in the future otherwise.

Comment (lasconic) [inline zerberus/sfz.cpp]: Delete the line if it's not needed.

File: zerberus/sfz.cpp [hunk 3]
```diff
@@ -472,44 +473,64 @@ void SfzRegion::readOp(const QString& b, const QString& data, SfzControl &c)
 bool ZInstrument::loadSfz(const QString& s)
       {
       _program = 0;
-      QFile f(s);
-      if (!f.open(QIODevice::ReadOnly)) {
+      QFile* f = new QFile(s);
+      if (!f->open(QIODevice::ReadOnly)) {
             qDebug("ZInstrument: cannot load %s", qPrintable(s));
```
Comment (lasconic) [inline zerberus/sfz.cpp]: delete f;

Comment (lasconic) [inline zerberus/sfz.cpp]: delete f;

File: zerberus/sfz.cpp [hunk 4]
```diff
@@ -472,44 +473,64 @@ void SfzRegion::readOp(const QString& b, const QString& data, SfzControl &c)
...
+
+      while (!f->atEnd()) {
+            QByteArray ba = f->readLine();
+            zerberus->setLoadProgress(((qreal)f->pos() * 100) / total);
             ba = ba.simplified();
 
             if (ba.isEmpty() || ba.startsWith("//"))
                   continue;
             if (zerberus->loadWasCanceled())
                   return false;
```
Comment (lasconic) [inline zerberus/sfz.cpp]: what about f?

File: zerberus/sfz.cpp (hunk 1 above)
Comment (lasconic) [inline zerberus/sfz.cpp]: delete f; ?

Comment (hpfmn) [inline zerberus/sfz.cpp]: here better not (there could be more on the stack)

Comment (hpfmn) [inline zerberus/sfz.cpp]: or better something that actually cleans the stack

File: zerberus/sfz.cpp [hunk 5]
```diff
@@ -16,6 +16,7 @@
 #include <QFileInfo>
 #include <QStringList>
 #include <sndfile.h>
+#include <stack>
```
Comment (lasconic) [inline zerberus/sfz.cpp]: the stack is not used right?

File: zerberus/sfz.cpp [hunk 6]
```diff
@@ -465,76 +465,146 @@ void SfzRegion::readOp(const QString& b, const QString& data, SfzControl &c)
...
+            if (curLine.startsWith("#define")) {
+                  QStringList define = curLine.split(" ");
+                  foundWithSpaces = findWithSpaces.match(curLine);
+                  if (define.size() == 3)
+                        c.defines.insert(std::pair<QString, QString>(define[1], define[2]));
+                  else if(foundWithSpaces.hasMatch())
+                        c.defines.insert(std::pair<QString, QString>(define[1], foundWithSpaces.captured(1)));
+                  fileContents.removeAt(idx);
+                  }
+            else if (curLine.startsWith("#include")) {
```
Comment (lasconic) [inline zerberus/sfz.cpp]: 3 levels #include will not work right?

Comment (hpfmn) [inline zerberus/sfz.cpp]: They should work. As you can see the contents of the file are inserted after the current line -> so if they contain another #include they should be inserted there etc.

Comment (hpfmn) [inline zerberus/sfz.cpp]: As said on IRC - first line of a new file and first line after #define was not considered by the preprocessor - should work now!

Comment (lasconic) [inline zerberus/sfz.cpp]: OK. That will do a good mtest. Merging this.

**Pull request merged**